### ISSUE TITLE
Parallel ParagraphVectors inference crashes the JVM

### ISSUE THREAD

**Issue by oliverdain** [deeplearning4j/nd4j#1637]
I have some Kotlin code which run the following on a large data set:

    val inferred = doc2Vec.inferVector(it)
    val label = doc2Vec.nearestLabels(it, 1).first()
    val labelVec = doc2Vec.lookupTable.vector(label)
    val similarity = Transforms.cosineSim(inferred, labelVec)

I used to run this in a single thread processing one data item at a time. That worked fine. I just parallelized by splitting the data set into `n` chunks and then running the above in `n` thread. When I did that things started to crash immediately, and not just throwing an  exception: it crashes the whole JVM with a `SIGSEGV`:

`[libnd4jcpu.so+0x10df2d] NativeOps::execAggregateFloat(void**, int, float**, int, int**, int, int*, int, int**, int, float*, int)+0x27cd`

**Comment by treo** [comment]
Can you please provide the full crashlog.

Also can you tell us, if you use multiple doc2vec model instances or are you reusing a single one?

**Comment by oliverdain** [comment]
It's a single doc2vec instance. Will get you the crash log shortly. I need to recreate it first which take a bit.

**Comment by treo** [comment]
I guess that may be the problem here. The model instances are not thread safe if I remember correctly, so you are probably seeing the crash because of this.

**Comment by oliverdain** [comment]
Oh! I asked in Gitter if they were safe a while back and I thought I was told that they are.

Would this be the right way to make a copy of the trained model:

        ParagraphVectors vectors = new ParagraphVectors.Builder(trained.getConfiguration())
                .vocabCache(trained.getVocab())
                .lookupTable(trained.getLookupTable())
                .resetModel(false)
                .build();

Stole that from the deserialization code...

**Comment by oliverdain** [comment]
Sounds like it's not relevant anymore, but just for posterity, here's the full exception:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f51cf7aba9b, pid=6716, tid=0x00007f51bbeb6700
#
# JRE version: OpenJDK Runtime Environment (8.0_121-b13) (build 1.8.0_121-b13)
# Java VM: OpenJDK 64-Bit Server VM (25.121-b13 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libnd4jcpu.so+0x10da9b]  NativeOps::execAggregateFloat(void**, int, float**, int, int**, int, int*, int, int**, int, float*, int)+0x233b
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /mnt/yarn/usercache/hadoop/appcache/application_1486587262859_0002/container_1486587262859_0002_01_000002/hs_err_pid6716.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by treo** [comment]
> /mnt/yarn/usercache/hadoop/appcache/application_1486587262859_0002/container_1486587262859_0002_01_000002/hs_err_pid6716.log

The contents of this file is what I was actually looking for :)

And lets get @agibsonccc and @raver119 in here, so they can give you a definitive answer on how it should be done.

As for copying the trained model, I'm not sure if that will work correctly or not, as it could end up with the same vocabCache and lookup table instance and I don't know about their thread safety.

**Comment by oliverdain** [comment]
@treo when my job died Yarn oh-so-helpfully cleaned up that log file. If you need it I can create it again but it sounds like the issue was just me using a non-thread-safe class in a threaded environment so it's probably not necessary. Let me know if you would like that file and I'd be happy to make things crash again.

**Comment by oliverdain** [comment]
Update: I just saw this same crash in the serial version. It's MUCH less common when things are running serially (have trained dozens of models serially without issue while the parallel version always crashes pretty quickly). So somehow this seems to be not just related to parallelization though that appears to make it much more likely.

**Comment by saudet** [comment]
@oliverdain It sounds like we might have to run this through Valgrind to figure out what's going on:
https://github.com/deeplearning4j/libnd4j/wiki/Debugging-libnd4j

**Comment by treo** [comment]
@oliverdain also, for the serial version, can you please get the crash log file? It contains a lot more information about the crash.

**Comment by oliverdain** [comment]
@treo shoot! Should have thought to grab the log when it happened. Unfortunately the logs are gone now (it was a spot instance in an EMR cluster and my bid was too low so it got reclaimed). If it happens again I'll be sure to grab that log file.

**Comment by raver119** [comment]
Sorry, i'm a bit sick for last few days, so my presence is limited here atm.

Problem with paravec inference - it wasn't really ever tested in spark/flink/whatever environment, mostly because inference is internally parallel. Like all other ops, because dl4j w2v/d2v is 99% native this days.

Algorithmically it looks like this:
1. You get text document, and pass it to inferVector() method
2. It gets tokenized using your tokenizer & preprocessor. This happens in SINGLE thread.
3. Collection<VocabWord> gets passed for real inferVector() method, and op batches will be created & issued as single native op call,  which will be applied in parallel.

So, obviously, if you want to speed up things as workaround - you need to tokenize your documents manually in parallel threads, and then pass those to infer vector. Optionally we might provide some queue-based method for batched inference with callbacks, but it's a bit out of scope for us really.

P.s. yep, @ monday i'll check what's wrong with parallel execution. I have few suspects.

**Comment by raver119** [comment]
At this moment i'm unable to reproduce crash mentioned above. I'm rolling through LabelAwareIterator & batched inference, and things just work so far.

**Comment by raver119** [comment]
I've implemented batched inference & I've eliminated bottleneck in parallel inference execution, and now i have 100% utilization on all 36 cores. Crash still not reproducible for me.

**Comment by oliverdain** [comment]
@raver119 thanks for doing that! Have you pushed to Maven/jcenter or do I need to build locally to get the goodness?

Have been working on other stuff for the last few days. When I get back to this will try to get more info/logs if I can reproduce the crash.

FWIW, we do have some data that contains things like just a single word or a single symbol mixed into our (messy) data set. I know that when I run locally that causes ParagraphVectors to throw an NPE. Not the same crash but perhaps if you try very short texts or texts that tokenize to the empty-set you'll be able to reproduce the crash.

**Comment by raver119** [comment]
Nope, not merged yet. Going to merge tomorrow after additional validation tests.

And thanks for additional info. I'll try to add those edge cases to tests and will see what happens.

**Comment by raver119** [comment]
btw, if you have that NPE stack trace around - could you please post it here? That would be very helpful.

**Comment by oliverdain** [comment]
@raver119  sorry, should have said that I already filed a bug on the NPE and it's been fixed: https://github.com/deeplearning4j/deeplearning4j/issues/2773

**Comment by raver119** [comment]
Ah, i've fixed that, right :)

Thanks anyway :)

**Comment by raver119** [comment]
Closing this issue for now, since it's not reproducible for me + batched inference was added.

Please file new issue if you'll manage to reproduce this issue + will have a crash og :)
